docs/design.md

```diff
@@ -0,0 +1,126 @@
+# Design of image update automation
+
+There are three parts to the design, that operate independently, and
+in sympathy with one another:
+
+ - the job runner (type UpdateJob)
+ - the image metadata reflector (types ImageRepository, ImagePolicy)
+ - the automation controller (type ImageAutomation)
+
+Some tooling comes alongside these parts:
+
+ - the `tk-image` command-line tool lets you create the resource
```

Comment:
I'm gesturing at an extension mechanism similar to git, whereby if `tk-image` is in the `$PATH`, tk will invoke it to handle the subcommand `tk image`.

Reply:
Sounds good, thanks for the clarification. Might be worthwhile to mention that idea in the doc.

Reply:
The toolkit CLI uses `<verb> <kind> <name> <args>`, I personally don't like the idea of plugins, it would break the current UX. Images should fit nicely into our current design `tk create/get/reconcile/delete/suspend/resume image repository|policy` same as sources.

Reply:
I don't care too much about operating as an extension. The trade-off in the other direction is that all the command-line toolery is centralised in `tk`, which must be changed every time there's a new type or system.

Reply:
The toolkit repo is where everything comes together, we have all the components installers, CRDs and APIs being assembled, tested and released under a single version. Having tk depend and use all the current APIs was a good exercise to deal with breaking changes and interoperability between them, for example `tk reconcile kustomization --with-source`. If each components comes with it's own CLI, the trade-off would be on the UX side, things like `--with-source` and other cross-apis commands wouldn't be possible.

Reply:
> If each component comes with it's own CLI, the trade-off would be on the UX side, things like `--with-source` and other cross-api commands wouldn't be possible.

This isn't really accurate as stated. It's possible, for example, to have things that _depend_ on source-controller types have an option `--with-source`, without them being part of `tk`. Assuming there is _some_ layering to the individual components -- e.g., the image stuff here depends on a source-controller type, but nothing in `tk` depends on the image stuff -- there can be tools that follow the same user interface and don't lose anything by being separate.

But I think you're going for another point, which is something like "putting everything in tk makes it obvious when the APIs are not interoperable". No argument here -- but I'd be careful not to make every piece need every other piece in order to work.